### PR TITLE
Add test for libvirtd crash regression bugs

### PR DESCRIPTION
**Pull request by Hao-Liu**
This PR depends on https://github.com/autotest/virt-test/pull/1948 and https://github.com/autotest/virt-test/pull/1947

This test runs various regression tests and check whether libvirt
daemon crashes.

Signed-off-by: Hao Liu hliu@redhat.com

### CODE REVIEW

File: libvirt/tests/src/daemon/crash_regression.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+    vm_name = vm.name
+    vm_xml = VMXML.new_from_inactive_dumpxml(vm_name)
+    vm_xml_backup = vm_xml.copy()
+    save_path = os.path.join(data_dir.get_tmp_dir(), 'tmp.save')
+    try:
+        pm_xml = VMPMXML()
+        pm_xml.mem_enabled = 'yes'
+        vm_xml.pm = pm_xml
+        vm_xml.sync()
+        VMXML.set_agent_channel(vm.name)
```
Comment (yumingfei) [inline libvirt/tests/src/daemon/crash_regression.py]: This function is not used in right format, maybe it has been modified by other guys.
```
    def set_agent_channel(self, src_path=None,
                          tgt_name='org.qemu.guest_agent.0',
                          ignore_exist=False):
        """
        Add a channel for guest agent if non exists.

        :param src_path: Source path of the channel
        :param tgt_name: Target name of the channel
        :param ignore_exist: Whether add a channel even if another already exists.
        """
```

File: libvirt/tests/src/daemon/crash_regression.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+def run_kill_virsh_while_managedsave(params, libvirtd, vm):
+    """
+    Kill virsh process when doing managedsave.
+    """
+    thread = threading.Thread(
+        target=virsh.managedsave,
+        args=(vm.name,)
+    )
+    thread.start()
+    utils.run('pkill virsh')
```
Comment (yumingfei) [inline libvirt/tests/src/daemon/crash_regression.py]: This case hanged at one round of 100.I have to execute several "pkill virsh" manually.
Log:
```
09:58:35 DEBUG| Setting ignore_status to True.
09:58:35 DEBUG| Running 'systemctl status libvirtd.service'
09:58:35 DEBUG| Stopping libvirtd service
09:58:35 DEBUG| Setting ignore_status to True.
09:58:35 DEBUG| Running 'systemctl stop libvirtd.service'
09:58:36 DEBUG| Waiting for gdb inferior libvirtd to start
09:58:36 DEBUG| gdb is Calling back start
09:58:36 DEBUG| Waiting for libvirtd to work
09:58:36 DEBUG| Running 'virsh list'
09:58:38 WARNI| run process timeout (2) fired on: virsh list
09:58:40 DEBUG| Running 'virsh list'
09:58:42 WARNI| run process timeout (2) fired on: virsh list
09:58:44 DEBUG| Running 'virsh list'
09:58:46 WARNI| run process timeout (2) fired on: virsh list
09:58:47 DEBUG| (address cache) DHCP lease OK: 52:54:00:2b:14:b6 --> 192.168.122.19
09:58:48 DEBUG| Running 'virsh list'
09:58:48 DEBUG| Running 'pkill virsh'
09:58:48 DEBUG| Running 'pkill virsh'
09:58:48 DEBUG| Running 'pkill virsh'
09:58:48 DEBUG| Running 'pkill virsh'
09:58:48 DEBUG| Running 'pkill virsh'
09:58:48 DEBUG| Running 'pkill virsh'
09:58:48 DEBUG| Running 'pkill virsh'
09:58:48 DEBUG| Running 'pkill virsh'
09:58:48 DEBUG| Running 'pkill virsh'
09:58:51 DEBUG| gdb is Calling back stop
09:58:51 DEBUG| gdb is Calling back signal
10:11:40 DEBUG| (address cache) DHCP lease OK: 8c:89:a5:b8:96:60 --> 10.167.226.81
***Hang up HERE*** # Go on after I execute several "pkill virsh".
10:24:18 DEBUG| Running 'pkill virsh'
10:24:18 DEBUG| Running 'pkill virsh'
10:24:18 DEBUG| Running 'pkill virsh'
10:24:18 DEBUG| Running 'pkill virsh'
10:24:18 DEBUG| Running 'pkill virsh'
10:24:18 DEBUG| Running 'pkill virsh'
10:24:18 DEBUG| Running 'pkill virsh'
10:24:18 DEBUG| Running 'pkill virsh'
```

File: libvirt/tests/src/daemon/crash_regression.py (hunk 1 above)
Comment (Hao-Liu) [inline libvirt/tests/src/daemon/crash_regression.py]: Well, It seems "the other guys" is me. I change that method in https://github.com/autotest/virt-test/pull/1961 after I wrote this test. I will change this now.

**Comment by Hao-Liu** [comment]
@yumingfei, I've solve the guest agent issue.

For the `pkill virsh` issue, I can't reproduce your result. But considering your log and method to resolve, I made following assumption:
* Because some unknown cause, your TERM signal emitted by `pkill virsh` command can't stop `virsh managedsave` process, but turn it into a locking status.
* Another one or more TERM signals can kill the process.
* Since existing `virsh managedsave` sessions hang and after sessions count exceed 10, new sessions
stop to start.
* After killed these process manually, test continues.

I changed be kill virsh session with signal KILL. and give a warning when `pkill` fails. I hope these changes can pass the test.

For the virsh hang issue, I can't debug that since I can't reproduce it. Maybe you can help check it out.

**Comment by yumingfei** [comment]
Yep, It runs good now.

**Comment by yumingfei** [comment]
Log:
```
16:17:48 DEBUG| Running 'pkill -9 virsh'
16:17:48 DEBUG| Running 'pkill -9 virsh'
16:17:48 DEBUG| Running 'pkill -9 virsh'
16:17:48 DEBUG| Running 'pkill -9 virsh'
16:17:48 DEBUG| Running 'pkill -9 virsh'
16:17:48 WARNI| Failed to kill virsh:
* Command: 
    pkill -9 virsh
Exit status: 1
Duration: 0.00796103477478

16:17:48 DEBUG| Running 'pkill -9 virsh'
16:17:48 DEBUG| Running 'pkill -9 virsh'
16:17:48 DEBUG| Running 'pkill -9 virsh'
16:17:48 DEBUG| Running 'pkill -9 virsh'
```
If it should be like this, I can merge it now. Please check it again.

**Comment by Hao-Liu** [comment]
It is not part of this test. But I don't thinks it is what it should be. It means the `virsh managedsave` process, which should not complete in less than 1s, is not found. Assuming it is caused by the managedsave thread have a latency to start, it should take more time to wait for virsh managedsave command to fully complete.
I'll update it to wait the managedsave thread to be started before kill it.

**Comment by Hao-Liu** [comment]
Found the cause of fail to kill: It is caused by the state of VM will be "shut off" if daemon managedsave is done. check and start VM before each loop is added.

**Comment by yumingfei** [comment]
OK, I got the reason for one of failure:
It failed to set selinux label for vm during managesaving, so I think it's useful to output of managedsave when pkill fails.
```
[root@domain haoliu_virt]# virsh managedsave f20_fyu
error: Failed to save domain f20_fyu state
error: unable to set security context 'system_u:object_r:svirt_image_t' on fd 26: Invalid argument
```

Then I met another problem. It often failed because of inappropriate time to pkill.
But it always got "Killed" when I did managedsave manually. So this needs to be improved.

BTW, they are all for case crash_regression.kill_virsh_while_managedsave.

**Comment by yumingfei** [comment]
Haaaa, It's funny, testcase run well after following changes, JFYI:
```
diff --git a/libvirt/tests/src/daemon/crash_regression.py b/libvirt/tests/src/daemon/crash_regression.py
index 36335b1..c7d2603 100644
--- a/libvirt/tests/src/daemon/crash_regression.py
+++ b/libvirt/tests/src/daemon/crash_regression.py
@@ -101,11 +101,15 @@ def run_kill_virsh_while_managedsave(params, libvirtd, vm):
     """
     if not vm.is_alive():
         vm.start()
+    utils.wait_for(lambda: vm.state().strip() in ["running"], timeout=5)
+
     thread = threading.Thread(
         target=virsh.managedsave,
         args=(vm.name,)
     )
     thread.start()
+    import time
+    time.sleep(1)
     try:
         res = utils.run('pkill -9 virsh', ignore_status=True)
         if res.exit_status:
```

BTW, I think you should set a timeout for this case to avoid to hang on at somewhere when libvirtd is down. I still met it recently.

**Comment by will-Do** [comment]
Let's merge this patch, any update, please raise a new pull request

**Pull request merged**